src/slotmap.js

```diff
@@ -274,4 +274,54 @@ function slotMapOf(token) {
     return map;
 }
 
-module.exports = { slotMapOf, packValues };
+/**
+ * Represents the value in the persistent storage.
+ * Each token has its own SlotMap. Any value can be assigned to this storage
+ * ad-hoc. It can be used for dynamically determined keys needed to be persistent.
+ */
+class PersistentStorageMap {
+    constructor() {
+        /** @type {Map<string, SlotMap>} */
+        this._map = new Map();
+    }
+
+    /**
+     * @param {string} tokenId
+     * @param {number} blockNumber
+     */
+    _init(tokenId, blockNumber) {
+        const key = `${tokenId}:${blockNumber}`;
+        const initialized = this._map.get(key) || new SlotMap();
+        if (!this._map.has(key)) {
+            this._map.set(key, initialized);
+        }
+        return initialized;
```

Comment:
to avoid the double guard

```suggestion
        let slotMap = this._map.get(key);
        if (slotMap === undefined) {
            slotMap = new SlotMap();
            this._map.set(key, slotMap);
        }        
        return slotMap;
```